### DISCUSSION TITLE
Gauge Invariance

### DISCUSSION THREAD

**Discussion by gj19866** [idaholab/moose#26022] [Q&A Modules: General] [answered]
Hi there,

I am wondering how MOOSE handles gauge invariance. When solving a set of problems that has gauge variance should the user select a predefined gauge to work within, or will gauge invariance be handled by the test function?
And subsequently, should the user define a gauge, will this affect the results/convergence of the kernels due to some interaction with the test function?

This might be one that @cticenhour might know the answer to.

Any help would be much appreciated!

Many thanks,
Gillian

**Comment by cticenhour** [comment, marked as answer]
We don't currently have a model example that applies a gauge and solves for a vector potential in the MOOSE electromagnetics module, as Maxwell's equations are manipulated into coupled wave equations and solved directly for the electric and magnetic vector fields. Adding a Coulomb gauge, in particular, is something that is on my roadmap for the coming months of development for a research project, but isn't ready yet. 

The gauge invariance isn't handled by the test function. We use a first-order Nedelec vector basis from libMesh for most problems, and the recent introduction of Raviart-Thomas elements in libMesh is something I also want to explore. So, if you wanted to attempt to define a gauge in your problem, the test function should not affect the results or convergence (at least in a way specific to the gauge, I believe). @lindsayad or @roystgnr may be able to give some insight on the libMesh test functions and correct me if I am wrong here.

**Reply by gj19866** [reply]
Ok great, thank you very much for the help!

I will keep an eye out for the addition of a Coulomb gauge, and I will impose my own gauge for now.

All the best,
Gillian

**Reply by lindsayad** [reply]
Without spending too much time re-learning gauge invariance, I would need to see the PDEs you want to solve

**Reply by roystgnr** [reply]
Disclaimer: I know basically nothing about gauge invariance.  It might be wise to stop reading this comment after that sentence.

Is the idea here that we're solving for a divergence-free vector field like the magnetic field B, and ∇·B=0 implies that there exists a potential-vector field A such that B=∇×A, so we'd like to be able to just solve for A directly and get a strongly divergence-free B, but the problem is that A is only unique up to some transformation (like adding ∇Ω for some smooth Ω) that changes A but not B?

If you're solving a problem like that in 2D, you can use a variable defined with one of the libMesh C¹ scalar-valued elements (`CLOUGH` works on general grids; `HERMITE` is more efficient if you have a cartesian mesh), treat its values as if they were magnitudes of a vector field in the z direction, take the curl of that field to get a divergence-free vector field in the xy plane, and the kernel of that operation is just the constant fields so you can add a single extra residual term to pin the value at one point to get a well-defined potential.

If you're solving a problem like that in 3D, you could still use three `HERMITE` variables to give you components of a C¹ vector field, and take the curl of that to get a divergence-free 3D vector field ... but now the kernel isn't a simple one-dimensional thing, it's dimensionality is something like a third of your degrees of freedom.  Maybe there's some efficient way to constrain that and keep going, but when I looked at that problem during my PhD I was advised to find something easier instead.
